[WebXR][GTK][WPE] Implement PlatformXrCoordinator::getPrimaryDeviceInfo()

svillar opened the pull request:
#### 19bf33d025b917636e0decc5f0aba9f21ff063a1
<pre>
[WebXR][GTK][WPE] Implement PlatformXrCoordinator::getPrimaryDeviceInfo()
<a href="https://bugs.webkit.org/show_bug.cgi?id=294843">https://bugs.webkit.org/show_bug.cgi?id=294843</a>

Reviewed by Carlos Garcia Campos.

The PlatformXrCoordinator is one of the most important object in the UIProcess
related to XR. Among other things it&apos;s responsible for:
* retrieving info and capabilities of the XR device/system
* start/stop XR sessions
* schedule frame rendering

For GTK &amp; WPE ports that coordinator is implemented using OpenXR. This PR
implements the method that retrieves the device capabilities as a first step.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webxr/webxr_availability.http.sub-expected.txt: Added.
* LayoutTests/platform/gtk/TestExpectations: Unskipped tests.
* LayoutTests/platform/gtk/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/feature-policy/reporting/xr-reporting.https-expected.txt:
* LayoutTests/platform/wpe/TestExpectations: Unskipped tests.
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.cpp: Added.
(PlatformXR::Instance::singleton):
(PlatformXR::Instance::Instance):
(PlatformXR::Instance::enumerateImmersiveXRDevices):
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.h: Added.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::resetState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.cpp: Added.
(WebKit::OpenXRExtensions::create):
(WebKit::OpenXRExtensions::OpenXRExtensions):
(WebKit::OpenXRExtensions::loadMethods):
(WebKit::OpenXRExtensions::isExtensionSupported const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.h: Added.
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h: Added.
(WebKit::createOpenXRStruct):
(WebKit::checkXrResult):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp: Added.
(WebKit::OpenXRCoordinator::OpenXRCoordinator):
(WebKit::OpenXRCoordinator::~OpenXRCoordinator):
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::createInstance):
(WebKit::OpenXRCoordinator::recommendedResolution const):
(WebKit::OpenXRCoordinator::collectViewConfigurations):
(WebKit::OpenXRCoordinator::initializeSystem):
(WebKit::OpenXRCoordinator::initializeDevice):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h: Added.
* Source/WebKit/UIProcess/XR/openxr/PlatformXRSystemOpenXR.cpp: Added.
(WebKit::PlatformXRSystem::xrCoordinator):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/cmake/FindOpenXR.cmake:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/296610@main">https://commits.webkit.org/296610@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/512b3b9db220b2b39e3e53731a465259ef230ffd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109076 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114287 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59376 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111039 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29419 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37302 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82889 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112024 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23402 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98244 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63334 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22802 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16384 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58971 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101605 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92772 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117405 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107634 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36124 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91910 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36495 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94508 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91717 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23343 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36625 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14378 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31975 "Hash 512b3b9d for PR 47071 does not build (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36021 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41526 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131906 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35714 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35750 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39053 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37400 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on LayoutTests/platform/gtk/TestExpectations:
```diff
@@ -258,11 +258,12 @@ webkit.org/b/254733 inspector/canvas/recording-offscreen-webgl-snapshots.html [
-webkit.org/b/208988 webxr [ Skip ]
+webkit.org/b/208988 imported/w3c/web-platform-tests/webxr/webGLCanvasContext_create_xrcompatible.https.html [ Failure ]
+webkit.org/b/208988 imported/w3c/web-platform-tests/webxr/xrViewport_valid.https.html [ Failure ]
+webkit.org/b/208988 imported/w3c/web-platform-tests/webxr/xrWebGLLayer_viewports.https.html [ Failure ]
```
If expectations are exactly the same in gtk and wpe, use TestExpectations from glib dir instead

carlosgcampos commented on LayoutTests/platform/gtk/imported/w3c/web-platform-tests/webxr/webxr_availability.http.sub-expected.txt:
```diff
@@ -0,0 +1,4 @@
+
+PASS Test webxr not available in insecure context
+PASS Test webxr not available in secure context in insecure context
+
```
Same for tests, put the expected file in glib dir.

svillar replied:
:ok_hand:

carlosgcampos commented on Source/WebCore/platform/xr/openxr/PlatformXROpenXR.h:
```diff
@@ -0,0 +1,30 @@
+
+#pragma once
+
+#if ENABLE(WEBXR) && USE(EMPTYXR)
```
What's emptyxr?

svillar replied:
It's basically used for platforms where no XR SDK is available. It allows you to still run the tests for the WebXR API without having to have a platform implementation. See https://bugs.webkit.org/show_bug.cgi?id=225578

carlosgcampos commented on Source/WebCore/platform/xr/openxr/PlatformXROpenXR.cpp:
```diff
@@ -0,0 +1,69 @@
+#include "config.h"
+#include "PlatformXROpenXR.h"
+
+#if ENABLE(WEBXR) && USE(OPENXR)
```
I'm confused, this seems to be a copy of PlatformXRCocoa.mm but this condition is `#if ENABLE(WEBXR) && !USE(OPENXR)`

svillar replied:
Yeah that's because the current cocoa implementation uses !OPENXR to refer to themselves because the OpenXR implementation was the first one

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
```diff
@@ -0,0 +1,84 @@
+#include <epoxy/egl.h>
+#else
+#include <EGL/egl.h>
+#endif
```
This should not be included from headers if possible

svillar replied:
OK, the code is a bit more complex now, but they're moved to the source file.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -0,0 +1,185 @@
+    ASSERT(RunLoop::isMain());
+    ASSERT(m_instance == XR_NULL_HANDLE);
+
+    Vector<char *> extensions;
```
```suggestion
    Vector<const char*, 2> extensions;
```

svillar replied:
OK, but we'll add more extensions in future patches

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -0,0 +1,185 @@
+    XrInstanceCreateInfo createInfo = createOpenXRStruct<XrInstanceCreateInfo, XR_TYPE_INSTANCE_CREATE_INFO >();
+    createInfo.applicationInfo = { "WebKit", 1, "WebKit", 1, XR_CURRENT_API_VERSION };
+    createInfo.enabledApiLayerCount = 0;
+    createInfo.enabledExtensionCount = (uint32_t) extensions.size();
```
```suggestion
    createInfo.enabledExtensionCount = static_cast<uint32_t>(extensions.size());
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -0,0 +1,185 @@
+    CHECK_XRCMD(xrEnumerateViewConfigurationViews(m_instance, m_systemId, m_currentViewConfiguration, viewCount, &viewCount, views.mutableSpan().data()));
+
+    // OpenXR is very flexible wrt views resolution, but the current WebKit architecture expects a single resolution for all views.
+    return { (int) (viewCount * views.first().recommendedImageRectWidth), (int) views.first().recommendedImageRectHeight };
```
Use c++ casts here too

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -0,0 +1,185 @@
+    ASSERT(RunLoop::isMain());
+    ASSERT(m_viewConfigurations.size());
+
+    uint32_t viewCount;
```
```suggestion
    uint32_t viewCount = 0;
```

carlosgcampos replied:
I don't know if initialization is needed, but it's done in other cases.

svillar replied:
Not needed at all.

carlosgcampos commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -407,6 +408,19 @@ if (USE_WOFF2)
+if (ENABLE_WEBXR)
+    if (NOT ENABLE_GAMEPAD)
+        message(FATAL_ERROR "Gamepad is required to be enabled for WebXR support.")
+    endif ()
```
I think we could use `WEBKIT_OPTION_DEPEND`

carlosgcampos commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -407,6 +408,19 @@ if (USE_WOFF2)
+        message(FATAL_ERROR "OpenXR is required to enable WebXR support.")
+    endif ()
+    SET_AND_EXPOSE_TO_BUILD(USE_OPENXR ${OpenXR_FOUND})
+    SET_AND_EXPOSE_TO_BUILD(XR_USE_PLATFORM_EGL TRUE)
+    SET_AND_EXPOSE_TO_BUILD(XR_USE_GRAPHICS_API_OPENGL_ES TRUE)
```
Why do we need these if they are always TRUE?

svillar replied:
First of all these are defined in openxr includes, so we need to set them to true to be able to use certain types/methods which otherwise would be disabled, it isn't a WebKit guard.

Apart from that other platforms, for example Android, will define different ones.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.cpp:
```diff
@@ -0,0 +1,93 @@
+
+std::unique_ptr<OpenXRExtensions> OpenXRExtensions::create()
+{
+    uint32_t extensionCount { 0 };
```
I guess it's not needed here either?

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -0,0 +1,185 @@
+    ASSERT(RunLoop::isMain());
+    ASSERT(m_instance != XR_NULL_HANDLE);
+
+    uint32_t viewConfigurationCount = 0;
```
Nor here

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
```diff
@@ -0,0 +1,62 @@
+    XrInstance m_instance { XR_NULL_HANDLE };
+    XrSystemId m_systemId { XR_NULL_SYSTEM_ID };
+    Vector<XrViewConfigurationType> m_viewConfigurations;
+    XrViewConfigurationType m_currentViewConfiguration;
```
Shouldn't this be initialized?

svillar replied:
We could but it does not really make much sense in this case because it is something provided by the runtime and there is no value we could assume as "default"

svillar committed:
[WebXR][GTK][WPE] Implement PlatformXrCoordinator::getPrimaryDeviceInfo()
https://bugs.webkit.org/show_bug.cgi?id=294843

Reviewed by Carlos Garcia Campos.

The PlatformXrCoordinator is one of the most important object in the UIProcess
related to XR. Among other things it's responsible for:
* retrieving info and capabilities of the XR device/system
* start/stop XR sessions
* schedule frame rendering

For GTK & WPE ports that coordinator is implemented using OpenXR. This PR
implements the method that retrieves the device capabilities as a first step.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webxr/webxr_availability.http.sub-expected.txt: Added.
* LayoutTests/platform/gtk/TestExpectations: Unskipped tests.
* LayoutTests/platform/gtk/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/feature-policy/reporting/xr-reporting.https-expected.txt:
* LayoutTests/platform/wpe/TestExpectations: Unskipped tests.
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.cpp: Added.
(PlatformXR::Instance::singleton):
(PlatformXR::Instance::Instance):
(PlatformXR::Instance::enumerateImmersiveXRDevices):
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.h: Added.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::resetState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.cpp: Added.
(WebKit::OpenXRExtensions::create):
(WebKit::OpenXRExtensions::OpenXRExtensions):
(WebKit::OpenXRExtensions::loadMethods):
(WebKit::OpenXRExtensions::isExtensionSupported const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.h: Added.
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h: Added.
(WebKit::createOpenXRStruct):
(WebKit::checkXrResult):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp: Added.
(WebKit::OpenXRCoordinator::OpenXRCoordinator):
(WebKit::OpenXRCoordinator::~OpenXRCoordinator):
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::createInstance):
(WebKit::OpenXRCoordinator::recommendedResolution const):
(WebKit::OpenXRCoordinator::collectViewConfigurations):
(WebKit::OpenXRCoordinator::initializeSystem):
(WebKit::OpenXRCoordinator::initializeDevice):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h: Added.
* Source/WebKit/UIProcess/XR/openxr/PlatformXRSystemOpenXR.cpp: Added.
(WebKit::PlatformXRSystem::xrCoordinator):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/cmake/FindOpenXR.cmake:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/296610@main
